src/Microsoft.Health.Dicom.Api/Web/LazyMultipartReadOnlyStream.cs

```diff
@@ -0,0 +1,227 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+
+namespace Microsoft.Health.Dicom.Api.Web
+{
+    // inspired by
+    // https://github.com/microsoft/referencesource/blob/master/System/net/System/Net/Http/MultipartContent.cs
+    // https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs
+#pragma warning disable CA1844 // Provide memory-based overrides of async methods when subclassing 'Stream'
+    public class LazyMultipartReadOnlyStream : Stream
+#pragma warning restore CA1844 // Provide memory-based overrides of async methods when subclassing 'Stream'
+    {
+#pragma warning disable CA2213 // Disposable fields should be disposed, (disposed in ReadAsync)
+        private readonly IAsyncEnumerator<DicomStreamContent> _asyncEnumerator;
+#pragma warning restore CA2213 // Disposable fields should be disposed
+        private readonly int _bufferSize;
+        private const string Crlf = "\r\n";
+        private readonly Encoding _defaultHttpEncoding = Encoding.GetEncoding(28591);
+        private readonly string _boundary;
+        private readonly string _intermediateBoundary;
+        private const int KB = 1024;
+
+        private byte[] _buffer;
+        private DicomStreamContent _currentStreamContent;
+        private int _bufferPosition;
+        private int _bufferLength;
+        private bool _terminating;
+
+        public LazyMultipartReadOnlyStream(
+            IAsyncEnumerable<DicomStreamContent> enumerableStreams,
+            string boundary,
+            int bufferSize,
+            CancellationToken cancellation)
+            : base()
+        {
+            EnsureArg.IsNotNull(enumerableStreams, nameof(enumerableStreams));
+            EnsureArg.IsNotEmptyOrWhiteSpace(boundary, nameof(boundary));
+            EnsureArg.IsGt(bufferSize, KB, nameof(bufferSize));
+
+            _asyncEnumerator = enumerableStreams.GetAsyncEnumerator(cancellation);
+            _bufferSize = bufferSize;
+            _buffer = ArrayPool<byte>.Shared.Rent(bufferSize);
```

Comment:
Would it make sense to defer renting until a call to `ReadAsync`?

Reply:
Right pattern, but there is not much value add in our scenario